✨ ipam display count of address claims bound in condition

adobley opened the pull request:
<!--  Thanks for sending a pull request! Please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api-provider-vsphere/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other)
Here are some other tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."
-->

**What this PR does / why we need it**:
while there are unbound ip address claims the condition will show the count of bound to total claims, giving some feedback on how much work is left

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1689 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
when using AddressesFromPools the vSphereMachine status will reflect how many claims are awaiting addresses
```

srm09 commented on pkg/services/govmomi/service.go:
```diff
@@ -441,6 +425,61 @@ func (vms *VMService) reconcileIPAddresses(ctx *virtualMachineContext) (bool, er
+			var err error
+			ctx.Logger.V(5).Info("fetching IPAddressClaim", "name", ipAddrClaimKey.String())
+			if err = ctx.Client.Get(ctx, ipAddrClaimKey, ipAddrClaim); err != nil && !apierrors.IsNotFound(err) {
+				ctx.Logger.Error(err, "error fetching IPAddressClaim", "name", ipAddrClaimName)
+				return nil, err
```
even with the error we should check for the state of the other IPAddressClaim objects and aggregate any errors at the end of the for loop, construct an error and return.

tylerschultz replied:
While I agree with the spirit of this suggestion, I'm not sure I agree with this one specifically.

The prior step in this reconciliation loop does a find or create for the claims. 

In the current step of the loop, the controller is waiting for the claim to have an address bound. The case where the claim is not found is handled. The only reason this if condition would be hit is if there is an api error. Continuing on and attempting to look up other claims will likely fail too. Gathering up more api errors seemingly provides diminished value.

Unless you think it is important, we'll skip on this suggestion, and instead look for other places where we can apply the suggested strategy. 

WDYT?

adobley replied:
@srm09

srm09 reviewed (changes requested):
some linter fixes to be made as well.
/ok-to-test

tylerschultz commented:
FYI - We think we've covered the feedback, but we haven't run this code and tried it out. We'll let you know when we think we're ready for another round of review.

tylerschultz commented:
Update: we've deployed the changes and we are happy with what we're seeing. 
- The condition status shows "0 of 1..." in a nice way when the vm is waiting for IPs.
- The status messages format nicely when multiple IPs cause errors in the VSphereVM yaml.

One area of possible concern would be that these changes include newlines when multiple errors are logged. 

This PR should be good for another round of review.

tylerschultz commented:
/retest-required

adobley commented:
/retest

adobley commented:
We were seeing build errors fetching a version of vm-operator. We bumped that in another PR (#1732) and have rebased on that branch for now.

adobley commented:
Due to #1733 we have rebased onto #1734 until that gets merged so that the tests can run.

adobley commented:
> Due to #1733 we have rebased onto #1734 until that gets merged so that the tests can run.

Rebased back onto master now that #1734 is fixed. This should be good to go once everything passes.

adobley commented:
Rebased as it had been a few. I'll keep my eye on this but if everything passes we should be good to go again.

srm09 commented:
/approve
/lgtm
